Code/GraphMol/MolEnumerator/MolEnumerator.cpp

```diff
@@ -0,0 +1,149 @@
+//
+//  Copyright (C) 2020 Greg Landrum and T5 Informatics GmbH
+//
+//   @@ All Rights Reserved @@
+//  This file is part of the RDKit.
+//  The contents are covered by the terms of the BSD license
+//  which is included in the file license.txt, found at the root
+//  of the RDKit source tree.
+//
+#include "MolEnumerator.h"
+#include <RDGeneral/Exceptions.h>
+#include <GraphMol/FileParsers/MolSGroupParsing.h>
+
+namespace RDKit {
+namespace MolEnumerator {
+
+void PositionVariationOp::initFromMol(const ROMol &mol) {
+  dp_mol.reset(new ROMol(mol));
+  initFromMol();
+}
+void PositionVariationOp::initFromMol() {
+  d_variationPoints.clear();
+  if (!dp_mol) {
+    return;
+  }
+  for (const auto bond : dp_mol->bonds()) {
+    std::string endpts;
+    std::string attach;
+    if (bond->getPropIfPresent(common_properties::_MolFileBondEndPts, endpts) &&
+        bond->getPropIfPresent(common_properties::_MolFileBondAttach, attach) &&
+        attach == "ANY") {
+      const Atom *atom = bond->getBeginAtom();
+      if (atom->getAtomicNum() == 0) {
+        atom = bond->getEndAtom();
+        if (atom->getAtomicNum() == 0) {
+          throw ValueErrorException(
+              "position variation bond does not have connection to a "
+              "non-dummy "
+              "atom");
+        }
+      }
+      d_dummiesAtEachPoint.push_back(bond->getOtherAtomIdx(atom->getIdx()));
+      std::stringstream iss(endpts);
+      std::vector<unsigned int> oats =
+          RDKit::SGroupParsing::ParseV3000Array<unsigned int>(iss);
+      // decrement the indices and do error checking:
+      for (auto &oat : oats) {
+        if (oat == 0 || oat > dp_mol->getNumAtoms()) {
+          throw ValueErrorException("Bad variation point index");
+        }
+        --oat;
+      }
+      d_variationPoints.push_back(std::make_pair(atom->getIdx(), oats));
+    }
+  }
+}
+
+std::vector<size_t> PositionVariationOp::getVariationCounts() const {
+  std::vector<size_t> res(d_variationPoints.size());
+  std::transform(
+      d_variationPoints.begin(), d_variationPoints.end(), res.begin(),
+      [](std::pair<unsigned int, std::vector<unsigned int>> pr) -> size_t {
+        return pr.second.size();
+      });
+  return res;
+}
+
+ROMol *PositionVariationOp::operator()(const std::vector<size_t> &which) const {
```

Comment:
for new APIs, why don't we just to to std::unique_ptr<ROMol> instead of returning raw ROMol pointers.  This forces us to do proper ownership and memory management.

Reply:
If this is a step too far, that's ok with me, but I think it will happen at some point.

Reply:
Makes perfect sense to me. I will do the same thing with `copy()`

Reply:
One small bit of ugliness is that the only place we currently use `operator()` with these things - the `enumerate()` function - immediately transfers ownership to a smart pointer, but I think it still makes sense. The EnumerateOps could end up being useful elsewhere and this makes the API more explicit